### PR TITLE
rockchip: add support for nanopi r6s

### PR DESCRIPTION
**Pull request by BWhitten**
As we have now hit kernel version 6.6, which contained a lot of upstream commits for the rk3588 platform the back ported patches required for this board are quite small.

The resulting image can be put on an SD card and booted to experiment with.
Needed some errata turning on to boot, and a race condition in the clk of the mmc driver resolved to see the sd card.

Not been able to pull the MAC from the EEPROM yet, nor have I tried out the network interfaces hence the WIP tag

Shout out to https://github.com/Mr-3S/ for the kick start with uboot patches

### CODE REVIEW

**Comment by 1715173329** [comment]
tbh we still have a bench of works to do to bring up rk3588 support

> Not been able to pull the MAC from the EEPROM yet

either the EEPROM does not exist or it's empty. friendlyelec does not provide unique macaddr for this board.

**Comment by BWhitten** [comment]
Ahhh misled by their dts, yes the schematic says not mounted. I'll drop that.
image

**Comment by BWhitten** [comment]
Refreshed to main, u-boot needed some adjustments as they are migrating to upstream dts'

> Sorry for dumb question. If this bring support for r6s, does it also support orange pi 5 plus with rk3588 ?

This PR doesn't currently, if it's supported upstream someone with a board could try and add it.

**Comment by BWhitten** [comment]
I have dropped the patches for the board in u-boot as since 2024.07 there is a generic target which is good enough to boot the image from the SD card.
Also found that my ethernet interfaces were wrong, when I tried this live in my home setup.

@1715173329 you mentioned that there was a bunch of work to bring up rk3588 support. As this board is basically functional as a home router now, what is out standing in your view to be able to merge this basic support?

**Comment by 1715173329** [comment]
https://github.com/1715173329/openwrt-official/tree/master-rockchip-rk3588

I have created a wip branch that added basic rk3588 support. It needs rebase but I'm a bit lazy recently xD
Will update it these days.

**Comment by BWhitten** [comment]
We have been doing the same work then, this PR is re-based and ready to go if you want to add to it.

**Comment by 1715173329** [comment]
Your patchset missed driver updates as well as the most important cpu scaling support - guessing you will not want to run such a powerful CPU on just 816 MHz ;)

**Comment by BWhitten** [comment]
Still is able to download at > 100Mbps, but fair if I missed a commit upstream feel free to tack it on ;)

**Comment by 1715173329** [comment]
https://github.com/1715173329/openwrt-official/tree/master-rockchip-rk3588

I just rebased the branch. However `302-mmc-allow-probe-to-defer-if-clock-is-not-ready.patch` is required to bring up sdmmc, not sure what's the cause yet.

**Comment by BWhitten** [comment]
The cause is due to a race condition in the clock provider,, the sdmmc comes on before the clock is ready however the error doesn't get percolated up for the driver to check again later.

This race looks like it has been there forever and I need to upstream my patch for this.

File: target/linux/rockchip/patches-6.6/125-dt-bindings-firmware-arm-scmi-Document-assigned-cloc.patch [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  rk3588-rock-5b.dtb: scmi: protocol@14: Unevaluated properties are not allowed ('assigned-clock-rates', 'assigned-clocks' were unexpected)
+
+Add the missing properties.
+
+Signed-off-by: Cristian Ciocaltea <cristian.ciocaltea@collabora.com>
+---
+ Documentation/devicetree/bindings/firmware/arm,scmi.yaml | 3 +++
+ 1 file changed, 3 insertions(+)
+
+--- a/Documentation/devicetree/bindings/firmware/arm,scmi.yaml
```
Comment (1715173329) [inline target/linux/rockchip/patches-6.6/125-dt-bindings-firmware-arm-scmi-Document-assigned-cloc.patch:20]: This patch can be omitted

File: target/linux/rockchip/patches-6.6/126-dt-bindings-serial-snps-dw-apb-uart-Relax-dma-names-.patch [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+arch/riscv/boot/dts/allwinner/sunxi-d1s-t113.dtsi-  dma-names = "rx", "tx";
+
+Do not enforce the order of the dma-names items.
+
+Signed-off-by: Cristian Ciocaltea <cristian.ciocaltea@collabora.com>
+---
+ .../devicetree/bindings/serial/snps-dw-apb-uart.yaml | 12 ++++++++++++
+ 1 file changed, 12 insertions(+)
+
+--- a/Documentation/devicetree/bindings/serial/snps-dw-apb-uart.yaml
```
Comment (1715173329) [inline target/linux/rockchip/patches-6.6/126-dt-bindings-serial-snps-dw-apb-uart-Relax-dma-names-.patch:39]: And this

File: target/linux/rockchip/patches-6.6/155-add-rk806-regulator-config.patch [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+Signed-off-by: Lucas Tanure <lucas.tanure@collabora.com>
+Signed-off-by: Sebastian Reichel <sebastian.reichel@collabora.com>
+---
+ .../dts/rockchip/rk3588-rk806-single.dtsi     | 181 ++++++++++++++++++
+ 1 file changed, 181 insertions(+)
+ create mode 100644 arch/arm64/boot/dts/rockchip/rk3588-rk806-single.dtsi
+
+--- /dev/null
++++ b/arch/arm64/boot/dts/rockchip/rk3588-rk806-single.dtsi
```
Comment (1715173329) [inline target/linux/rockchip/patches-6.6/155-add-rk806-regulator-config.patch:16]: Upstream does not use it instead write it each dts. Should be removed.

File: target/linux/rockchip/armv8/base-files/etc/board.d/02_network [hunk 4]
```diff
@@ -24,6 +24,9 @@ rockchip_setup_interfaces()
 	friendlyarm,nanopi-r5s)
 		ucidef_set_interfaces_lan_wan 'eth1 eth2' 'eth0'
 		;;
+	friendlyarm,nanopi-r6s)
```
Comment (1715173329) [inline target/linux/rockchip/armv8/base-files/etc/board.d/02_network:27]: Please also update `target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity`

File: target/linux/rockchip/patches-6.6/125-dt-bindings-firmware-arm-scmi-Document-assigned-cloc.patch (hunk 1 above)
Comment (BWhitten) [reply target/linux/rockchip/patches-6.6/125-dt-bindings-firmware-arm-scmi-Document-assigned-cloc.patch]: ok

File: target/linux/rockchip/patches-6.6/126-dt-bindings-serial-snps-dw-apb-uart-Relax-dma-names-.patch (hunk 2 above)
Comment (BWhitten) [reply target/linux/rockchip/patches-6.6/126-dt-bindings-serial-snps-dw-apb-uart-Relax-dma-names-.patch]: ok

File: target/linux/rockchip/patches-6.6/155-add-rk806-regulator-config.patch (hunk 3 above)
Comment (BWhitten) [reply target/linux/rockchip/patches-6.6/155-add-rk806-regulator-config.patch]: ok

File: target/linux/rockchip/armv8/base-files/etc/board.d/02_network (hunk 4 above)
Comment (BWhitten) [reply target/linux/rockchip/armv8/base-files/etc/board.d/02_network:27]: done

**Comment by BWhitten** [comment]
> BTW uboot has merged support for R6S, so you can back port it instead:

done

**Comment by BWhitten** [comment]
@1715173329 Updated to make formalities check happy and confirm that I have run tested this on my unit at home
Removing [WIP,RFC]

**Comment by 1715173329** [comment]
commit `uboot-rockchip: backport upstream dts sync` is really huge. Backporting [rk3588s-nanopi-r6s.dts](https://github.com/u-boot/u-boot/blob/b7d5ce05b1b52136af09cf34541c6a91079a4b86/dts/upstream/src/arm64/rockchip/rk3588s-nanopi-r6s.dts) is enough IMO.

**Comment by BWhitten** [comment]
> commit `uboot-rockchip: backport upstream dts sync` is really huge. Backporting [rk3588s-nanopi-r6s.dts](https://github.com/u-boot/u-boot/blob/b7d5ce05b1b52136af09cf34541c6a91079a4b86/dts/upstream/src/arm64/rockchip/rk3588s-nanopi-r6s.dts) is enough IMO.

It is a direct import of the commit upstream, when we move to the next uboot version it should be clear that it can be dropped entirely? I can butcher it if we really want, @hauke preference?

**Comment by antnyfls** [comment]
> It is a direct import of the commit upstream, when we move to the next uboot version it should be clear that it can be dropped entirely? I can butcher it if we really want

@BWhitten No need  everything is in the sync patch is 2.5MB only the R6S DTS, I backported for the R6C https://github.com/antnyfls/openwrt/commit/b2736faca670f2068e379b8722affd697ab7e5bb

**Comment by BWhitten** [comment]
Okay trimmed down the uboot patch to just the r6s, @1715173329  good to go?

**Review by 1715173329 — APPROVED** [review]

**Review by Mohsen7s — APPROVED** [review]

**Comment by hauke** [comment]
> > commit `uboot-rockchip: backport upstream dts sync` is really huge. Backporting [rk3588s-nanopi-r6s.dts](https://github.com/u-boot/u-boot/blob/b7d5ce05b1b52136af09cf34541c6a91079a4b86/dts/upstream/src/arm64/rockchip/rk3588s-nanopi-r6s.dts) is enough IMO.
> 
> It is a direct import of the commit upstream, when we move to the next uboot version it should be clear that it can be dropped entirely? I can butcher it if we really want, @hauke preference?

Normally I prefer backporting the complete patch, but 2.5MB is really big, so this partially port is better.

File: target/linux/rockchip/patches-6.6/302-mmc-allow-probe-to-defer-if-clock-is-not-ready.patch
Comment (hauke) [inline target/linux/rockchip/patches-6.6/302-mmc-allow-probe-to-defer-if-clock-is-not-ready.patch:1]: Is someone sending this patch to upstream Linux?
This looks fine to me.

Comment (BWhitten) [reply target/linux/rockchip/patches-6.6/302-mmc-allow-probe-to-defer-if-clock-is-not-ready.patch:1]: I'm planning on sending once I remember the ins and outs of the mailing lists. Thanks!

**Comment by gabefair** [comment]
For those looking, the openwrt tech page can be found here:

https://openwrt.org/toh/hwdata/friendlyarm/friendlyarm_nanopi_r6s

**Pull request merged**